pkg/collector/corechecks/cluster/ksm/kubernetes_state.go

```diff
@@ -219,22 +224,33 @@ type KSMConfig struct {
 	PodCollectionMode podCollectionMode `yaml:"pod_collection_mode"`
 }
 
+type resourceDiscoveryInfo struct {
+	resources  []*v1.APIResourceList
+	custom     customResources
+	collectors []string
+}
+
 // KSMCheck wraps the config and the metric stores needed to run the check
 type KSMCheck struct {
 	core.CheckBase
-	agentConfig          model.Config
-	instance             *KSMConfig
-	allStores            [][]cache.Store
-	telemetry            *telemetryCache
-	cancel               context.CancelFunc
-	isCLCRunner          bool
-	isRunningOnNodeAgent bool
-	clusterNameTagValue  string
-	clusterNameRFC1123   string
-	metricNamesMapper    map[string]string
-	metricAggregators    map[string]metricAggregator
-	metricTransformers   map[string]metricTransformerFunc
-	metadataMetricsRegex *regexp.Regexp
+	agentConfig           model.Config
+	instance              *KSMConfig
+	allStores             [][]cache.Store
+	telemetry             *telemetryCache
+	cancel                context.CancelFunc
+	isCLCRunner           bool
+	isRunningOnNodeAgent  bool
+	clusterNameTagValue   string
+	clusterNameRFC1123    string
+	metricNamesMapper     map[string]string
+	metricAggregators     map[string]metricAggregator
+	metricTransformers    map[string]metricTransformerFunc
+	metadataMetricsRegex  *regexp.Regexp
+	resourceDiscoveryInfo *resourceDiscoveryInfo
+	retrier               retry.Retrier
+	initRetry             sync.Once
+	runConfigureOnce      sync.Once
```

Comment:
I think it's actually not useful, if `configureError != nil`, you always return `configureError`, there's no multiple goroutines running a single instance check so that should be fine.